hal/src/rtc/embassy.rs

```diff
@@ -0,0 +1,158 @@
+//! This is a driver for embassy-time
+//! It must be instantiated with embassy_time!()
+//! which will configure the chip's RTC
+
+use core::cell::RefCell;
+use core::task::Waker;
+
+use crate::pac::{Interrupt, Rtc, NVIC};
+use crate::rtc::modes::{
+    mode0::{Compare0, RtcMode0},
+    RtcMode,
+};
+use atsamd_hal_macros::hal_cfg;
+use critical_section::{CriticalSection, Mutex};
+use embassy_time_driver::Driver;
+use embassy_time_queue_utils::Queue;
+
+/// Used internally by the embassy time driver.
+/// You shouldn't need this
+pub struct EmbassyBackend {
+    queue: Mutex<RefCell<Queue>>,
+}
+
+impl EmbassyBackend {
+    #[allow(clippy::new_without_default)]
+    pub const fn new() -> Self {
+        Self {
+            queue: Mutex::new(RefCell::new(Queue::new())),
+        }
+    }
+
+    fn set_alarm(&self, _cs: &CriticalSection, at: u64, rtc: &Rtc) -> bool {
+        // Embassy uses u64::MAX as a "no upcoming interrupt" sentinel
+        let at = match u32::try_from(at) {
+            Ok(at) => at,
+            _ if at == u64::MAX => u32::MAX,
+            Err(_) => return false,
+        };
+
+        RtcMode0::set_compare(rtc, 0, at);
+        true
```

Comment:
i am experiencing race condition in this implementation where it is possible for the RTC counter to advance past `at` while COMP0 is being set

the [[embassy timer queue implementation for stm32 microcontrollers]](https://github.com/embassy-rs/embassy/blob/6186d111a5c150946ee5b7e9e68d987a38c1a463/embassy-stm32/src/time_driver.rs#L449-L514) tries to prevent this from happening by avoiding setting alarms for timestamps in the very near future: